### PR TITLE
[RLlib] Rllib 2.0 policy/model API

### PR DESCRIPTION
**Pull request by kouroshHakha**
## Why are these changes needed?

This PR is an API draft for us to validate whether our design would solve the difficulty of customization in RLlib and ease the on-boarding process while maintaining the capability of RLlib. The code draft will evolve under `rllib2` folder.  

An RL `Algorithm` is going to be responsible for "when" and "what" of an algorithm not the "hows. The "how"s will be delegated to one or more `TrainerUnit`s. (In case of heterogenous multi-agent RL we will have multiple TrainerUnit objects within one Algorithm, e.g. PPOAgents + DQNAgents.

Each `TrainerUnit` (aka old `Policy`) will be responsible for constructing the model, optimizer, and loss function for training all the neural networks.

Under `TrainerUnit` we use `RLModule` which is a thin wrapper around all the neural networks that are needed for training and inference of an RL algorithm. Each algorithm creates its own `RLModule` because only the algorithm would know whether it needs a Q_function or not, no-one else. 

Inside `RLModule` user can call the model_catalog or provide customization for common RL pieces like `Pi`, `ValueFunction`, `QFunction`, .etc. 

`RLModule` is also used, for sample collection during both evaluation or environment rollout for online data collection.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib2/algorithms/ppo/torch/unit_trainer.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+from typing import Dict
+import torch
+from torch.optim import Optimizer, Adam
+
+from rllib2.data.sample_batch import SampleBatch
+from rllib2.core.torch.torch_unit_trainer import UnitTrainer
+from rllib2.core.torch.torch_rl_module import TorchRLModule
+
+
+class PPOTorchUnitTrainer(UnitTrainer):
```
Comment (sven1977) [inline rllib2/algorithms/ppo/torch/unit_trainer.py:11]: Should we just call these e.g. `PPOTorch`? That would simplify the nomenclature a bit AND make it very clear to the users that everything Torch specific that PPO "brings with it" (e.g. loss) should go into this class.

I do like "UnitTrainer" (now that we disambiguated and no longer have RLlib Trainers), but also ok with "TrainUnit" (shorter and kind of sounds better).

File: rllib2/algorithms/ppo/torch/unit_trainer.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        # ? how do we make a generic customizable TorchRLModule?
+        pass
+
+    def make_optimizer(self) -> Dict[str, Optimizer]:
+        return {
+            'pi': Adam(self.model.pi.parameters(), lr=self.config['pi_lr']),
+            'vf': Adam(self.model.vf.parameters(), lr=self.config['vf_lr'])
+        }
+
+    def loss(
```
Comment (sven1977) [inline rllib2/algorithms/ppo/torch/unit_trainer.py:23]: `compute_loss` or `compute_losses`

File: rllib2/core/algorithm.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import rllib.agents
+
+
+class Algorithm(rllib.agents.Trainer):
```
Comment (sven1977) [inline rllib2/core/algorithm.py:4]: `rllib.algorithms.algorithm.Algorithm`

File: rllib2/core/torch/torch_rl_module.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    pi_out_prev: Optional[PiOutput] = None
+    vf: Optional[VFOutput] = None
+
+
+class PPOTorchRLModule(TorchRLModule):
+
+    def __init__(self, config: PPORLModuleConfig):
+        super().__init__(config)
+
+        self.pi: Pi = model_catalog.make_pi(
```
Comment (sven1977) [inline rllib2/core/torch/torch_rl_module.py:100]: This would support both "default Pis" and custom ones, correct?

Depending on the values in the config.

Default:
config.pi.custom_module=None
config.pi.fcnet_hiddens = [128, 128, 128]

Custom:
config.pi.custom_module=[MyCustomPiClass]
config.pi.ctor_kwarg1 = [64, 64]
config.pi.ctor_kwarg2 = "relu"

**Comment by kouroshHakha** [comment]
don't close :)

**Comment by kouroshHakha** [comment]
don't close it :)
